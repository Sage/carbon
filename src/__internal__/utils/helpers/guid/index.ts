/**
 * Generates a random guid, useful for creating unique IDs.
 */
export default (): string => {
  function s4(): string {
    return Math.floor((1 + Math.random()) * 0x10000)
      .toString(16)
      .substring(1);
  }
  return `${s4()}${s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
};
