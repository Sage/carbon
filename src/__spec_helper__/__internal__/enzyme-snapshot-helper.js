import toJson from "enzyme-to-json";

function snapshotWithoutProps(enzymeWrapper, propNames) {
  let propNameList = propNames;

  if (!Array.isArray(propNames)) {
    propNameList = [propNames];
  }

  const snapshotOptions = {
    map: (json) => {
      propNameList.forEach((prop) => {
        if (json.props[prop]) {
          json.props[prop] = `[ ${prop} object ]`;
        }
      });

      return json;
    },
  };

  return toJson(enzymeWrapper, snapshotOptions);
}

function noThemeSnapshot(enzymeWrapper) {
  return snapshotWithoutProps(enzymeWrapper, "theme");
}

export { snapshotWithoutProps, noThemeSnapshot };
