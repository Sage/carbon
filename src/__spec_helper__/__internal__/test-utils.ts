/* eslint-disable jest/no-conditional-expect */
/* eslint-disable jest/no-identical-title */
/* eslint-disable jest/no-export */
import { render } from "@testing-library/react";
import { sprintf } from "sprintf-js";
import {
  LayoutProps,
  FlexboxProps,
  BackgroundProps,
  PositionProps,
  GridProps,
} from "styled-system";

import { space } from "../../style/themes/base/base-theme.config";
import { mockMatchMedia } from "../mock-match-media";

const isUpper = (char: string) => char.toUpperCase() === char;
const humpToDash = (acc: string, char: string) =>
  `${acc}${isUpper(char) ? `-${char.toLowerCase()}` : char}`;

const toCSSCase = (str: string) => {
  return str.split("").reduce(humpToDash, "");
};

const assertStyleMatch = (
  styleSpec: { [key: string]: string | number | undefined },
  component: HTMLElement,
  opts?: jest.Options,
) => {
  Object.entries(styleSpec).forEach(([attr, value]) => {
    expect(component).toHaveStyleRule(toCSSCase(attr), value, opts);
  });
};

const makeArrayKeys = (n: number) => [...Array(n).keys()];

const marginProps = [
  ["m", "margin"],
  ["ml", "marginLeft"],
  ["mr", "marginRight"],
  ["mt", "marginTop"],
  ["mb", "marginBottom"],
  ["mx", "marginLeft"],
  ["mx", "marginRight"],
  ["my", "marginTop"],
  ["my", "marginBottom"],
] as const;

type MarginProps = {
  [K in (typeof marginProps)[number][0]]?: string | number;
};

const paddingProps = [
  ["p", "padding"],
  ["pl", "paddingLeft"],
  ["pr", "paddingRight"],
  ["pt", "paddingTop"],
  ["pb", "paddingBottom"],
  ["px", "paddingLeft"],
  ["px", "paddingRight"],
  ["py", "paddingTop"],
  ["py", "paddingBottom"],
] as const;

type PaddingProps = {
  [K in (typeof paddingProps)[number][0]]?: string | number;
};

const colorProps = [
  ["color", "color", "#CCCCCC"],
  ["bg", "background-color", "#FFFFFF"],
  ["opacity", "opacity", "0.5"],
] as const;

const widthProps = ["width", "width", "200px"] as const;

const heightProps = ["height", "height", "200px"] as const;

const layoutProps = [
  widthProps,
  heightProps,
  ["minWidth", "min-width", "120px"],
  ["maxWidth", "max-width", "120px"],
  ["minHeight", "min-height", "120px"],
  ["maxHeight", "max-height", "120px"],
  ["size", "width", "120px"],
  ["size", "height", "120px"],
  ["display", "display", "inline-block"],
  ["verticalAlign", "vertical-align", "baseline"],
  ["overflow", "overflow", "hidden"],
  ["overflowX", "overflow-x", "hidden"],
  ["overflowY", "overflow-y", "hidden"],
] as const;

const flexBoxProps = [
  ["alignItems", "alignItems", "center"],
  ["alignContent", "alignContent", "center"],
  ["justifyItems", "justifyItems", "center"],
  ["justifyContent", "justifyContent", "center"],
  ["flexWrap", "flexWrap", "wrap"],
  ["flexDirection", "flexDirection", "row-reverse"],
  ["flex", "flex", "1"],
  ["flexGrow", "flexGrow", "1"],
  ["flexShrink", "flexShrink", "1"],
  ["flexBasis", "flexBasis", "100px"],
  ["justifySelf", "justifySelf", "center"],
  ["alignSelf", "alignSelf", "center"],
  ["order", "order", "1"],
] as const;

const gridProps = [
  ["gridColumn", "gridColumn", "1 / span 2"],
  ["gridRow", "gridRow", "1 / span 2"],
  ["gridArea", "gridArea", "myArea"],
  ["gridAutoFlow", "gridAutoFlow", "column"],
  ["gridAutoRows", "gridAutoRows", "150px"],
  ["gridAutoColumns", "gridAutoColumns", "50px"],
  ["gridTemplateRows", "gridTemplateRows", "100px 300px"],
  ["gridTemplateColumns", "gridTemplateColumns", "auto auto auto auto"],
  ["gridTemplateAreas", "gridTemplateAreas", "myArea myArea . . ."],
] as const;

const backgroundProps = [
  [
    "background",
    "background",
    "lightblue url('foo.jpg') no-repeat fixed center",
  ],
  ["backgroundImage", "background-image", "url(foo.jpg)"],
  ["backgroundSize", "background-size", "center"],
  ["backgroundRepeat", "background-repeat", "no-repeat"],
] as const;

const positionProps = [
  ["top", "0px"],
  ["bottom", "0px"],
  ["right", "0px"],
  ["left", "0px"],
  ["position", "fixed"],
  ["position", "sticky"],
  ["position", "absolute"],
  ["position", "static"],
  ["position", "relative"],
] as const;

export const getDefaultValue = (value?: string | number) => {
  const spaceArrayLength = space.length - 1;
  if (value === undefined) return value;
  const parsedValue = +value;
  if (typeof value === "string" && parsedValue > spaceArrayLength) {
    return `${value}px`;
  }
  if (parsedValue <= spaceArrayLength) {
    return space[parsedValue];
  }
  if (parsedValue > spaceArrayLength) {
    return `${parsedValue * 8}px`;
  }
  return value;
};

const testStyledSystemGrid = (
  component: (gridProperties?: GridProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(gridProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component(props));

        assertStyleMatch({ [propName]: value }, elementQuery());
      });
    },
  );
};

const testStyledSystemBackground = (
  component: (backgroundProperties?: BackgroundProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(backgroundProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));

        assertStyleMatch({ [styledSystemProp]: value }, elementQuery());
      });
    },
  );
};

const testStyledSystemPosition = (
  component: (positionProperties?: PositionProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(positionProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, value) => {
      it(`then ${styledSystemProp} should have been set correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));

        assertStyleMatch({ [styledSystemProp]: value }, elementQuery());
      });
    },
  );
};

// this util will catch that a console output occurred without polluting the output when running the unit tests
const expectConsoleOutput = (
  message: string,
  type: "warn" | "error" = "error",
) => {
  if (!message) {
    throw new Error(`no ${type} message provided`);
  }

  expect.assertions(1);

  const consoleType = global.console[type];
  let consoleArgs: string[];

  jest.spyOn(global.console, type).mockImplementation((...args) => {
    if (!args.length) return;

    const msg = args.join(" ");
    const params = args.slice(1, args.length);

    if (sprintf(msg, ...params).includes(message)) {
      consoleArgs = args;
      return;
    }

    consoleType(...args);
  });

  return () => {
    if (consoleArgs) {
      // eslint-disable-next-line no-console
      expect(console[type]).toHaveBeenCalledWith(...consoleArgs);
    }

    global.console[type] = consoleType;
  };
};

const testStyledSystemMargin = (
  component: (spacingProps?: MarginProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: MarginProps,
  assertOpts?: jest.Options,
) => {
  describe("default props", () => {
    let StyleElement: HTMLElement;

    beforeAll(() => {
      render(component({ ...defaults }));
      StyleElement = elementQuery();
    });

    it("should set the correct margins", () => {
      let margin;
      let marginLeft;
      let marginRight;
      let marginTop;
      let marginBottom;

      if (defaults) {
        margin = getDefaultValue(defaults.m);
        marginLeft = getDefaultValue(defaults.ml || defaults.mx);
        marginRight = getDefaultValue(defaults.mr || defaults.mx);
        marginTop = getDefaultValue(defaults.mt || defaults.my);
        marginBottom = getDefaultValue(defaults.mb || defaults.my);

        assertStyleMatch(
          {
            margin,
            marginLeft,
            marginRight,
            marginTop,
            marginBottom,
          },
          StyleElement,
          assertOpts,
        );
      } else {
        expect(StyleElement).not.toHaveStyleRule("marginLeft");
        expect(StyleElement).not.toHaveStyleRule("marginRight");
        expect(StyleElement).not.toHaveStyleRule("marginTop");
        expect(StyleElement).not.toHaveStyleRule("marginBottom");
        expect(StyleElement).not.toHaveStyleRule("margin");
      }
    });
  });

  describe.each(marginProps)(
    'when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: 2 };
        render(component({ ...props }));

        assertStyleMatch(
          { [propName]: "var(--spacing200)" },
          elementQuery(),
          assertOpts,
        );
      });
    },
  );
};

const testStyledSystemPadding = (
  component: (spacingProps?: PaddingProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: PaddingProps,
  assertOpts?: jest.Options,
) => {
  describe("default props", () => {
    let StyleElement: HTMLElement;

    beforeAll(() => {
      render(component({ ...defaults }));
      StyleElement = elementQuery();
    });

    it("should set the correct paddings", () => {
      let padding;
      let paddingLeft;
      let paddingRight;
      let paddingTop;
      let paddingBottom;

      if (defaults) {
        padding = getDefaultValue(defaults.p);
        paddingLeft = getDefaultValue(defaults.pl || defaults.px);
        paddingRight = getDefaultValue(defaults.pr || defaults.px);
        paddingTop = getDefaultValue(defaults.pt || defaults.py);
        paddingBottom = getDefaultValue(defaults.pb || defaults.py);

        assertStyleMatch(
          {
            padding,
            paddingLeft,
            paddingRight,
            paddingTop,
            paddingBottom,
          },
          StyleElement,
          assertOpts,
        );
      } else {
        expect(StyleElement).not.toHaveStyleRule("paddingLeft");
        expect(StyleElement).not.toHaveStyleRule("paddingRight");
        expect(StyleElement).not.toHaveStyleRule("paddingTop");
        expect(StyleElement).not.toHaveStyleRule("paddingBottom");
        expect(StyleElement).not.toHaveStyleRule("padding");
      }
    });
  });

  describe.each(paddingProps)(
    'when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: 2 };
        render(component({ ...props }));

        assertStyleMatch(
          { [propName]: "var(--spacing200)" },
          elementQuery(),
          assertOpts,
        );
      });
    },
  );
};

const testStyledSystemSpacing = (
  component: (spacingProps?: MarginProps | PaddingProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: MarginProps | PaddingProps,
  assertOpts?: jest.Options,
) => {
  testStyledSystemMargin(
    component,
    elementQuery,
    defaults as MarginProps,
    assertOpts,
  );
  testStyledSystemPadding(
    component,
    elementQuery,
    defaults as PaddingProps,
    assertOpts,
  );
};

const testStyledSystemColor = (
  // https://stackoverflow.com/questions/53711454/styled-system-props-typing-with-typescript
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  component: (colorProperties?: any) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(colorProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));
        const StyleElement = elementQuery();
        // Some props need to have camelcase so used toHaveStyleRule rather than assertStyleMatch
        expect(StyleElement).toHaveStyleRule(propName, value);
      });
    },
  );
};

const testStyledSystemWidth = (
  component: (widthProperties?: { width: string }) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe("when a width prop is specified using styled system props", () => {
    it("should set the width styling correctly", () => {
      const [styledSystemProp, propName, value] = widthProps;
      const props = { [styledSystemProp]: value };
      render(component({ ...props }));
      const StyleElement = elementQuery();
      expect(StyleElement).toHaveStyleRule(propName, value);
    });
  });
};

const testStyledSystemHeight = (
  component: (heightProperties?: { height: string }) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe("when a height prop is specified using styled system props", () => {
    it("should set the height styling correctly", () => {
      const [styledSystemProp, propName, value] = heightProps;
      const props = { [styledSystemProp]: value };
      render(component({ ...props }));
      const StyleElement = elementQuery();
      expect(StyleElement).toHaveStyleRule(propName, value);
    });
  });
};

const testStyledSystemLayout = (
  component: (layoutProperties?: LayoutProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(layoutProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));
        const StyleElement = elementQuery();
        // Some props need to have camelcase so used toHaveStyleRule rather than assertStyleMatch
        expect(StyleElement).toHaveStyleRule(propName, value);
      });
    },
  );
};

const testStyledSystemFlexBox = (
  component: (flexboxProperties?: FlexboxProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(flexBoxProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component(props));

        assertStyleMatch({ [propName]: value }, elementQuery());
      });
    },
  );
};

export {
  assertStyleMatch,
  toCSSCase,
  makeArrayKeys,
  mockMatchMedia,
  testStyledSystemColor,
  testStyledSystemLayout,
  testStyledSystemFlexBox,
  testStyledSystemGrid,
  testStyledSystemBackground,
  testStyledSystemPosition,
  expectConsoleOutput,
  testStyledSystemSpacing,
  testStyledSystemMargin,
  testStyledSystemPadding,
  testStyledSystemWidth,
  testStyledSystemHeight,
};
