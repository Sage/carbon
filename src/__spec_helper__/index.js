import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { setup } from './mock-match-media';

setup();
Enzyme.configure({ adapter: new Adapter() });
