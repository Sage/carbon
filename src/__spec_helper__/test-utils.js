import { mount } from 'enzyme';

import { carbonThemeList } from '../style/themes';
import { mockMatchMedia } from './mock-match-media';

const isUpper = char => char.toUpperCase() === char;
const humpToDash = (acc, char) => `${acc}${isUpper(char) ? `-${char.toLowerCase()}` : char}`;

const toCSSCase = (str) => {
  return str.split('').reduce(humpToDash, '');
};

const assertStyleMatch = (styleSpec, component, opts) => {
  Object.entries(styleSpec).forEach(([attr, value]) => {
    expect(component).toHaveStyleRule(toCSSCase(attr), value, opts);
  });
};

const makeArrayKeys = n => [...Array(n).keys()];

const dispatchKeyPress = (code) => {
  const ev = new KeyboardEvent('keydown', { which: code });
  document.dispatchEvent(ev);
};

const keyMap = {
  UpArrow: '38',
  DownArrow: '40',
  RightArrow: '39',
  LeftArrow: '37',
  Enter: '13',
  Tab: '9',
  Space: '32',
  Escape: '27',
  End: '35',
  Home: '36',
  D: '68',
  E: '69',
  P: '80',
  Z: '90',
  1: '49'
};

const repeat = action => (n = 1) => makeArrayKeys(n).forEach(() => action());

const keyboard = Object.keys(keyMap).reduce((acc, key) => {
  acc[`press${key}`] = () => repeat(dispatchKeyPress(keyMap[key]));
  return acc;
}, {});

// Build an object of Enzyme simulate helpers
// e.g. simulate.keydown.pressTab(target, { shiftKey: true })
// e.g. simulate.keydown.pressEscape(target)
const keydown = Object.keys(keyMap).reduce((acc, key) => {
  acc[`press${key}`] = (target, { shiftKey } = { shiftKey: false }) => {
    target.simulate('keydown', { shiftKey, key, which: parseInt(keyMap[key], 0) });
  };
  return acc;
}, {});

const simulate = {
  keydown
};

const listFrom = wrapper => wrapper.find('ul');
const childrenFrom = node => node.children();

const hoverList = wrapper => (item) => {
  childrenFrom(listFrom(wrapper)).at(item).simulate('mouseover');
};

const simulateEvent = eventName => wrapper => wrapper.simulate(eventName);
const click = simulateEvent('click');

const selectedItemOf = wrapper => wrapper.state().selectedItem;
const isUnique = (val, index, self) => self.indexOf(val) === index;
const isSelectableGiven = nonSelectables => i => !nonSelectables.includes(i);

const selectedItemReducer = method => wrapper => (acc, i) => {
  method(wrapper)(i);
  return [...acc, selectedItemOf(wrapper)];
};

const arraysEqual = (arr1, arr2) => arr1.sort().join(',') === arr2.sort().join(',');

const assertCorrectTraversal = method => expect => ({ num, nonSelectables = [] }) => (wrapper) => {
  const array = makeArrayKeys(num);
  const validIndexes = array.filter(isSelectableGiven(nonSelectables));


  const selectedItem = selectedItemOf(wrapper);
  const indexesThatWereSelected = array
    .reduce(selectedItemReducer(method)(wrapper), [selectedItem])
    .filter(isUnique);
  expect(arraysEqual(validIndexes, indexesThatWereSelected)).toBeTruthy();
};

const assertKeyboardTraversal = assertCorrectTraversal(() => keyboard.pressDownArrow)(expect);
const assertHoverTraversal = assertCorrectTraversal(wrapper => hoverList(wrapper))(expect);

const carbonThemesJestTable = carbonThemeList.map(theme => [theme.name, theme]);

const spacingProps = [
  ['m', 'margin'],
  ['ml', 'marginLeft'],
  ['mr', 'marginRight'],
  ['mt', 'marginTop'],
  ['mb', 'marginBottom'],
  ['mx', 'marginLeft'],
  ['mx', 'marginRight'],
  ['my', 'marginTop'],
  ['my', 'marginBottom'],
  ['p', 'padding'],
  ['pl', 'paddingLeft'],
  ['pr', 'paddingRight'],
  ['pt', 'paddingTop'],
  ['pb', 'paddingBottom'],
  ['px', 'paddingLeft'],
  ['px', 'paddingRight'],
  ['py', 'paddingTop'],
  ['py', 'paddingBottom']
];

const getDefaultValue = (value) => {
  if (typeof value === 'number') {
    return `${value * 8}px`;
  }
  return value;
};

const testStyledSystemSpacing = (component, defaults, styleContainer) => {
  describe('default props', () => {
    const wrapper = mount(component());
    const StyleElement = styleContainer ? styleContainer(wrapper) : wrapper;

    it('should set the correct margins', () => {
      let margin;
      let marginLeft;
      let marginRight;
      let marginTop;
      let marginBottom;

      if (defaults) {
        margin = getDefaultValue(defaults.m || undefined);
        marginLeft = getDefaultValue(defaults.ml || defaults.mx || undefined);
        marginRight = getDefaultValue(defaults.mr || defaults.mx || undefined);
        marginTop = getDefaultValue(defaults.mt || defaults.my || undefined);
        marginBottom = getDefaultValue(defaults.mb || defaults.my || undefined);

        expect(assertStyleMatch(
          {
            margin,
            marginLeft,
            marginRight,
            marginTop,
            marginBottom
          },
          StyleElement
        ));
      } else {
        expect(StyleElement).not.toHaveStyleRule('marginLeft');
        expect(StyleElement).not.toHaveStyleRule('marginRight');
        expect(StyleElement).not.toHaveStyleRule('marginTop');
        expect(StyleElement).not.toHaveStyleRule('marginBottom');
        expect(StyleElement).not.toHaveStyleRule('margin');
      }
    });

    it('should set the correct paddings', () => {
      let padding;
      let paddingLeft;
      let paddingRight;
      let paddingTop;
      let paddingBottom;

      if (defaults) {
        padding = getDefaultValue(defaults.p || undefined);
        paddingLeft = getDefaultValue(defaults.pl || defaults.px || undefined);
        paddingRight = getDefaultValue(defaults.pr || defaults.px || undefined);
        paddingTop = getDefaultValue(defaults.pt || defaults.py || undefined);
        paddingBottom = getDefaultValue(defaults.pb || defaults.py || undefined);

        expect(assertStyleMatch(
          {
            padding,
            paddingLeft,
            paddingRight,
            paddingTop,
            paddingBottom
          },
          StyleElement
        ));
      } else {
        expect(StyleElement).not.toHaveStyleRule('paddingLeft');
        expect(StyleElement).not.toHaveStyleRule('paddingRight');
        expect(StyleElement).not.toHaveStyleRule('paddingTop');
        expect(StyleElement).not.toHaveStyleRule('paddingBottom');
        expect(StyleElement).not.toHaveStyleRule('padding');
      }
    });
  });

  describe.each(spacingProps)('when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`then that ${propName} should have been set correctly`, () => {
        let wrapper = mount(component());

        const props = { [styledSystemProp]: 2 };
        wrapper = mount(component({ ...props }));

        expect(assertStyleMatch(
          { [propName]: '16px' },
          styleContainer ? styleContainer(wrapper) : wrapper
        ));
      });
    });
};

export {
  assertStyleMatch,
  toCSSCase,
  hoverList,
  selectedItemOf,
  childrenFrom,
  makeArrayKeys,
  keyboard,
  assertKeyboardTraversal,
  assertHoverTraversal,
  listFrom,
  click,
  simulate,
  carbonThemesJestTable,
  mockMatchMedia,
  testStyledSystemSpacing
};
