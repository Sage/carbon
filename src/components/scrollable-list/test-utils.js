import React from 'react';
import { makeArrayKeys } from '../../__spec_helper__/test-utils';
import ScrollableListItem from './scrollable-list-item.component';

const listItemReducer = ({ nonSelectables = [], customSelectables = [] }) => {
  // generate jsx for selectable list items, based on indexes in config
  return (acc, item, index) => {
    if (nonSelectables.includes(item)) return [...acc, <div />];
    if (customSelectables.includes(item)) return [...acc, <div isSelectable />];

    return [...acc, <ScrollableListItem id={ index } />];
  };
};

const renderListItems = (opts) => {
  return makeArrayKeys(opts.num).reduce(listItemReducer(opts), []);
};

export default renderListItems;
