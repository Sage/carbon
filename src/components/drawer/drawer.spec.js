import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount as enzymeMount, shallow } from 'enzyme';
import Drawer from './drawer.component';
import { assertStyleMatch } from '../../__spec_helper__/test-utils';
import guid from '../../utils/helpers/guid';
import {
  StyledDrawerSidebar,
  StyledDrawerContent,
  StyledDrawerChildren,
  StyledButton
} from './drawer.style';
import { noThemeSnapshot } from '../../__spec_helper__/enzyme-snapshot-helper';

jest.mock('../../utils/helpers/guid');
guid.mockImplementation(() => 'guid-123');

let container = null;

const defaultProps = {
  expandedWidth: '20%',
  animationDuration: '0.5s',
  sidebar: (
    <ul>
      <li>link a</li>
      <li>link b</li>
      <li>link c</li>
    </ul>
  )
};

const mount = (jsx) => {
  return enzymeMount(jsx, { attachTo: container });
};

const render = (props, renderer = mount) => {
  return renderer(
    <Drawer { ...props }>
      content body content body content body content body content body content body content body
    </Drawer>
  );
};

const getElements = (wrapper) => {
  const cw = wrapper;

  if (!cw) {
    return {};
  }

  return {
    drawer: cw.find(Drawer),
    sidebar: cw.find(StyledDrawerSidebar),
    content: cw.find(StyledDrawerContent),
    children: cw.find(StyledDrawerChildren),
    button: cw.find(StyledButton)
  };
};

describe('Drawer', () => {
  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    jest.useFakeTimers();
  });

  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  describe('uncontrolled', () => {
    it('matches snapshot', () => {
      const wrapper = render(defaultProps, shallow);
      expect(noThemeSnapshot(wrapper)).toMatchSnapshot();
    });

    it('cleans ups timers on unmount', () => {
      const wrapper = render();
      wrapper.unmount();
      expect(clearTimeout).toHaveBeenCalled();
    });

    it('correctly sets aria attribute', () => {
      const ariaLabel = 'test';
      const wrapper = render({ 'aria-label': ariaLabel });
      const { drawer } = getElements(wrapper);
      expect(drawer.prop('aria-label')).toBe(ariaLabel);
    });

    it('renders drawer component correctly', () => {
      const dataAttr = 'drawer';
      const wrapper = render({ 'data-component': dataAttr });
      const { drawer } = getElements(wrapper);
      expect(drawer.prop('data-component')).toBe(dataAttr);
    });

    it('Drawer Sidebar should render as expected', () => {
      const wrapper = render();
      const { sidebar } = getElements(wrapper);
      assertStyleMatch({
        overflow: 'auto'
      }, sidebar);
    });

    it('Drawer Content should render as expected', () => {
      const wrapper = render();
      const { children } = getElements(wrapper);
      assertStyleMatch({
        flex: '1',
        overflow: 'auto'
      }, children);
    });

    it('opens sidebar to specific width matching expandedWidth prop', () => {
      const expandedWidth = '50%';
      const wrapper = render({ expandedWidth });
      const { button, content } = getElements(wrapper);
      button.simulate('click');

      assertStyleMatch({
        width: expandedWidth
      }, content.childAt(0), { modifier: '&.open' });
    });

    it('one click on button opens drawer sidebar', () => {
      const wrapper = render();
      const { button } = getElements(wrapper);
      button.simulate('click');
      const { content } = getElements(wrapper);
      expect(content.childAt(0).hasClass('open')).toBeTruthy();
    });

    it('two clicks on button opens and then closes drawer sidebar', () => {
      const wrapper = render();
      let { button } = getElements(wrapper);
      button.simulate('click');
      let { content } = getElements(wrapper);
      expect(content.childAt(0).hasClass('open')).toBeTruthy();
      button = getElements(wrapper).button;
      button.simulate('click');
      content = getElements(wrapper).content;
      expect(content.childAt(0).hasClass('closed')).toBeTruthy();
    });

    it('sets class `open` on drawer that opened', () => {
      const wrapper = render();
      const { button } = getElements(wrapper);
      act(() => {
        button.simulate('click');
        jest.runAllTimers();
      });
      wrapper.update();
      const { content } = getElements(wrapper);
      expect(content.childAt(0).hasClass('open')).toBeTruthy();
    });

    it('sets class `closed` on drawer that closed', () => {
      const wrapper = render();
      const { button } = getElements(wrapper);
      act(() => {
        button.simulate('click');
        jest.runAllTimers();
      });
      wrapper.update();
      let { content } = getElements(wrapper);
      expect(content.childAt(0).hasClass('open')).toBeTruthy();

      act(() => {
        button.simulate('click');
        jest.runAllTimers();
      });
      wrapper.update();
      content = getElements(wrapper).content;
      expect(content.childAt(0).hasClass('closed')).toBeTruthy();
    });

    it('correctly renders opened drawer if defaultExpanded prop is true', () => {
      const defaultExpanded = true;
      const wrapper = render({ defaultExpanded });
      const { content } = getElements(wrapper);
      expect(content.childAt(0).hasClass('open')).toBeTruthy();
    });

    describe('invariant', () => {
      beforeEach(() => {
        jest.spyOn(global.console, 'error').mockImplementation(() => {});
      });

      afterEach(() => {
        global.console.error.mockReset();
      });

      it('throws if Drawer is changed from uncontrolled to controlled', () => {
        expect(() => {
          const wrapper = render({ expanded: undefined });
          wrapper.setProps({ expanded: true });
          wrapper.update();
        }).toThrow(
          'Drawer should not switch from uncontrolled to controlled (or vice versa). Decide between'
          + ' using a controlled or uncontrolled Drawer element for the lifetime of the component'
        );
      });
    });
  });

  describe('controlled', () => {
    it('sidebar is open when expanded prop is provided', () => {
      const wrapper = render({ expanded: true });
      const { content } = getElements(wrapper);
      assertStyleMatch({
        width: '40%'
      }, content.childAt(0), { modifier: '&.open' });
    });

    it('drawer changes to closed when button is clicked', () => {
      const wrapper = render({ expanded: true });
      const { button } = getElements(wrapper);
      button.simulate('click');
      const { content } = getElements(wrapper);

      assertStyleMatch({
        width: undefined
      }, content.childAt(0), { modifier: '&.closed' });
    });

    it('drawer changes to open when button is clicked', () => {
      const onChange = jest.fn();
      const wrapper = render({ expanded: false, onChange });
      const { button } = getElements(wrapper);
      button.simulate('click');
      const { content } = getElements(wrapper);

      assertStyleMatch({
        width: '40%'
      }, content.childAt(0), { modifier: '&.open' });
      expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('drawer opening sets timeout class', () => {
      const wrapper = render({ expanded: false, animationDuration: '500ms' });
      const { button } = getElements(wrapper);
      button.simulate('click');
      const { content } = getElements(wrapper);
      expect(content.childAt(0).hasClass('opening')).toBeTruthy();
    });

    it('sets `closing` class on drawer when close icon was clicked', () => {
      const wrapper = render({ expanded: true, animationDuration: '0.5s' });
      const { button } = getElements(wrapper);
      button.simulate('click');
      const { content } = getElements(wrapper);
      expect(content.childAt(0).hasClass('closing')).toBeTruthy();
    });

    it('sets animation speed to two seconds when string `numeric` value is given as a string', () => {
      const animationDuration = '2000';
      const wrapper = render({ expanded: true, animationDuration });
      const { button } = getElements(wrapper);
      button.simulate('click');
      const { content } = getElements(wrapper);
      expect(content.childAt(0).prop('animationDuration')).toBe(animationDuration);
    });

    it('sets animation speed to two seconds when string `ms` value is given as a string', () => {
      const animationDuration = '2000ms';
      const wrapper = render({ expanded: true, animationDuration });
      const { button } = getElements(wrapper);
      button.simulate('click');
      const { content } = getElements(wrapper);
      expect(content.childAt(0).prop('animationDuration')).toBe(animationDuration);
    });

    it('sets animation speed to two seconds when string `seconds` value is given as a string', () => {
      const animationDuration = '0.5s';
      const wrapper = render({ expanded: true, animationDuration });
      const { button } = getElements(wrapper);
      button.simulate('click');
      const { content } = getElements(wrapper);
      expect(content.childAt(0).prop('animationDuration')).toBe(animationDuration);
    });

    describe('invariant', () => {
      beforeEach(() => {
        jest.spyOn(global.console, 'error').mockImplementation(() => {});
      });

      afterEach(() => {
        global.console.error.mockReset();
      });

      it('throws if Drawer is changed from controlled to uncontrolled', () => {
        expect(() => {
          const wrapper = render({ expanded: true });
          wrapper.setProps({ expanded: undefined });
          wrapper.update();
        }).toThrow(
          'Drawer should not switch from uncontrolled to controlled (or vice versa). Decide between'
          + ' using a controlled or uncontrolled Drawer element for the lifetime of the component'
        );
      });
    });
  });
});
