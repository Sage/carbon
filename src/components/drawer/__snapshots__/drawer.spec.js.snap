// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer uncontrolled matches snapshot 1`] = `
<styled.div
  data-component="drawer"
  height="100%"
>
  <styled.div
    animationDuration="0.5s"
    aria-expanded="true"
    className="open"
    expandedWidth="20%"
    theme="[ theme object ]"
  >
    <Styled(styled.div)
      hasControls={false}
      id="DrawerSidebar_guid-123"
      isExpanded={true}
      overflowY="auto"
      role="navigation"
      scrollVariant="light"
      theme="[ theme object ]"
    >
      <ContextProvider
        value={
          Object {
            "isInSidebar": true,
          }
        }
      >
        <ul>
          <li>
            link a
          </li>
          <li>
            link b
          </li>
          <li>
            link c
          </li>
        </ul>
      </ContextProvider>
    </Styled(styled.div)>
  </styled.div>
  <styled.div>
    content body content body content body content body content body content body content body
  </styled.div>
</styled.div>
`;
